pkg/pki/pki.go

```diff
@@ -0,0 +1,123 @@
+ *                                                                            *
+ *                         Distributed Systems Runtime                        *
+ *                                                                            *
+ * -------------------------------------------------------------------------- *
+ *                                                                            *
+ *   Licensed under the Apache License, Version 2.0 (the "License");          *
+ *   you may not use this file except in compliance with the License.         *
+ *   You may obtain a copy of the License at                                  *
+ *                                                                            *
+ *       http://www.apache.org/licenses/LICENSE-2.0                           *
+ *                                                                            *
+ *   Unless required by applicable law or agreed to in writing, software      *
+ *   distributed under the License is distributed on an "AS IS" BASIS,        *
+ *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. *
+ *   See the License for the specific language governing permissions and      *
+ *   limitations under the License.                                           *
+ *                                                                            *
+\* -------------------------------------------------------------------------- */
+
+package pki
+
+import (
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/sha1"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/pem"
+	"fmt"
+	"math/big"
+	"os"
+	"time"
+)
+
+func CreateRootCA(dir string, domainName string) {
+	err := os.MkdirAll(dir, os.ModePerm)
+	if err != nil {
+		fmt.Println("Failed to create directory:", err)
+		os.Exit(1)
+	}
+
+	template := x509.Certificate{}
+	template.Subject = pkix.Name{
+		Organization:       []string{"Aurae"},
+		OrganizationalUnit: []string{"Runtime"},
+		StreetAddress:      []string{"aurae"},
+		Locality:           []string{"aurae"},
+		Country:            []string{"IS"},
+		CommonName:         domainName,
+	}
+
+	template.NotBefore = time.Now()
+	template.NotAfter = template.NotBefore.Add(24 * time.Hour * 9999)
+	// template.KeyUsage = x509.KeyUsageCertSign | x509.KeyUsageKeyEncipherment | x509.KeyUsageDigitalSignature | x509.KeyUsageCRLSign
+	// template.ExtKeyUsage = []x509.ExtKeyUsage{x509.ExtKeyUsageClientAuth, x509.ExtKeyUsageServerAuth}
+	template.IsCA = true
+	template.BasicConstraintsValid = true
+	template.DNSNames = []string{domainName}
+	priv, err := rsa.GenerateKey(rand.Reader, 2048)
+	if err != nil {
+		fmt.Println("Failed to generate private key:", err)
+		os.Exit(1)
+	}
+	pub := priv.PublicKey
+	serialNumberLimit := new(big.Int).Lsh(big.NewInt(1), 128)
+	template.SerialNumber, err = rand.Int(rand.Reader, serialNumberLimit)
+	if err != nil {
+		fmt.Println("Failed to generate serial number:", err)
```

Comment:
We should probably plumb through a more generic error handling mechanism so we aren't printing and exiting all over the place.

Reply:
I agree, will add proper error handling.